src/audio/pipeline/pipeline-params.c

```diff
@@ -162,8 +165,27 @@ static int pipeline_comp_hw_params(struct comp_dev *current,
 
 	/* Fetch hardware stream parameters from DAI component */
 	if (dev_comp_type(current) == SOF_COMP_DAI) {
-		ret = comp_dai_get_hw_params(current,
-					     &ppl_data->params->params, dir);
+#if CONFIG_IPC_MAJOR_3
+#ifdef CONFIG_ZEPHYR_NATIVE_DRIVERS
+		struct dai_data *dd;
+
+		/* dai-zephyr case */
+		dd = comp_get_drvdata(current);
+		ret = dai_common_get_hw_params(dd, current, &ppl_data->params->params, dir);
+#else
+		/* dai-legacy case */
+		if (current->drv->ops.dai_get_hw_params)
+			ret = current->drv->ops.dai_get_hw_params(current,
+								  &ppl_data->params->params, dir);
+#endif
+#elif CONFIG_IPC_MAJOR_4
+		struct copier_data *cd = comp_get_drvdata(current);
+		struct dai_data *dd = cd->dd[0];
+
+		ret = dai_common_get_hw_params(dd, current, &ppl_data->params->params, dir);
```

Comment:
Not specific to this commit, but with the recent copier / DAI integration work, I'm a bit concerned about mixing up of the layers. Now we have copier and DAI layers mixing in with the pipeline code. Is that an intermediate step or is it what we want eventually?

Reply:
I understand.

Before this change, the get hw params operation still looks very strange, it calls it by an inline ops driver interface, and this interface is not an standard for module adapter, so this is why this patch exists.

Regarding the change, due to we have to support multiple cases(otherwise, CI will report error)
looks a little redundant, maybe one day, ipc3 support was totally removed from SOF, then we may figure out a better way to implement this get hw params.

currently, seems this is the only way to remove this interface from copier.